### PR TITLE
Add vanilla megablox to MoE

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Add vanilla Megablox implementation (works well on single chip, not optimal sharding on multi hosts) to:
* further improve performance
* unblock Inference team's tests on Megablox (vipannalla@'s request)

# Test

* run `decode.py` got same outputs of matmul implementation - [link](https://pantheon.corp.google.com/kubernetes/service/us-central2/v4-bodaborg/default/ran-matmul-decode-t5/logs?e=13803378&mods=allow_workbench_image_override&project=tpu-prod-env-multipod). 

```
Input `[INST] I love to [/INST]` -> `That's great to hear! Could you please tell me what`
```
* run & add `forward_pass_logit_checker.py` test - [link](https://pantheon.corp.google.com/kubernetes/service/us-central2/v4-bodaborg/default/ran-matmul-decode-matmul-m2/logs?e=13803378&mods=allow_workbench_image_override&project=tpu-prod-env-multipod)

### CODE REVIEW

**Comment by RissyRan** [comment]
There is an issue on GPU with `jax.experimental.pallas.ops.tpu.megablox` import. I think this depends on JAX version. What could be a potential fix?

**Review by vipannalla — APPROVED** [review]
LGTM

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh [hunk 1]
```diff
@@ -34,6 +34,9 @@ python3 MaxText/tests/forward_pass_logit_checker.py  MaxText/configs/base.yml lo
 # Test whether the forward pass logits match the golden logits - matmul implementation
 python3 MaxText/tests/forward_pass_logit_checker.py  MaxText/configs/base.yml load_parameters_path=${MATMUL_SCANNED_CHECKPOINT} run_name=matmul_forward_pass_test per_device_batch_size=1 model_name=mixtral-8x7b tokenizer_path=gs://maxtext-external/mixtral-8x7B-v0.1-Instruct/tokenizer.mistral ici_tensor_parallelism=4 ici_fsdp_parallelism=16 max_prefill_predict_length=11 max_target_length=11 dataset_type=synthetic dtype=float32 --atol=3 --rtol=1 --token_size=4
 
+# Test whether the forward pass logits match the golden logits - megablox implementation
+python3 MaxText/tests/forward_pass_logit_checker.py MaxText/configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} load_parameters_path=${MATMUL_SCANNED_CHECKPOINT} run_name=megablox_forward_pass_test per_device_batch_size=1 model_name=mixtral-8x7b tokenizer_path=gs://maxtext-external/mixtral-8x7B-v0.1-Instruct/tokenizer.mistral ici_tensor_parallelism=4 ici_fsdp_parallelism=16 max_prefill_predict_length=11 max_target_length=11 dataset_type=synthetic dtype=float32 moe_matmul=True megablox=True --atol=3 --rtol=1 --token_size=4
```
Comment (gobbleturk) [inline end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh:38]: The atol and rtol are very high here. I missed this in the prior PR. Both the atol and rtol should be about 0.1, are the tests not passing for values near 0.1?

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -315,6 +326,69 @@ def generate_kernels(self, num_experts, base_emb_dim, mlp_dim):
     wo_kernel = jnp.asarray(wo_kernel, self.dtype)
     return w0_kernel, w1_kernel, wo_kernel
 
+  def permute(self, inputs, gate_logits, emb_dim):
+    inputs_2d = jnp.reshape(inputs, (-1, emb_dim))
+    weights, selected_experts = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
+    weights = jax.nn.softmax(weights.astype(self.weight_dtype), axis=-1).astype(self.dtype)
+    flatten_selected_experts = jnp.ravel(selected_experts)
+    sorted_selected_experts = jnp.argsort(flatten_selected_experts)
+    repeat_hidden_states = jnp.repeat(inputs_2d, self.num_experts_per_tok, axis=0)
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers/linears.py:335]: Can you explain (in the comments and also here) what the purpose of the repeat is? I cannot follow exactly what the next jnp.take line is doing - it looks like we are setting indeces=experts, e.g. the maximum index is only ~8 (the number of experts), so the majority of the repeat_hidden_states is unused (any row larger than the number of experts)

File: MaxText/layers/linears.py [hunk 3]
```diff
@@ -315,6 +326,69 @@ def generate_kernels(self, num_experts, base_emb_dim, mlp_dim):
...
+    flatten_weights = jnp.ravel(weights)
+    combined_output = jnp.multiply(unsort_output, flatten_weights[:, None])
+    groups = jnp.reshape(combined_output, (-1, self.num_experts_per_tok, combined_output.shape[1]))
+    return jnp.sum(groups, axis=1).reshape(inputs.shape).astype(self.dtype)
+  
+  def call_gmm(self, inputs, group_sizes, mlp_activation, w0_kernel, w1_kernel, wo_kernel):
+    @functools.partial(
+        shard_map.shard_map,
+        mesh=self.mesh,
+        in_specs=(
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:357]: Why are we choosing None for the sharding here? Can we add a comment?

**Review by gobbleturk — CHANGES_REQUESTED** [review]

File: MaxText/layers/linears.py [hunk 4]
```diff
@@ -315,6 +326,69 @@ def generate_kernels(self, num_experts, base_emb_dim, mlp_dim):
     wo_kernel = jnp.asarray(wo_kernel, self.dtype)
     return w0_kernel, w1_kernel, wo_kernel
 
+  def permute(self, inputs, gate_logits, emb_dim):
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:329]: Can you include a comment here for what this function does? I'm not sure if permute is the best name - maybe like get_inputs_for_experts?

File: MaxText/layers/linears.py [hunk 5]
```diff
@@ -315,6 +326,69 @@ def generate_kernels(self, num_experts, base_emb_dim, mlp_dim):
...
+    inputs_2d = jnp.reshape(inputs, (-1, emb_dim))
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:330]: Yes, correct. Will add a comment there.

File: MaxText/layers/linears.py [hunk 6]
```diff
@@ -315,6 +326,69 @@ def generate_kernels(self, num_experts, base_emb_dim, mlp_dim):
...
+    weights, selected_experts = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
+    weights = jax.nn.softmax(weights.astype(self.weight_dtype), axis=-1).astype(self.dtype)
+    flatten_selected_experts = jnp.ravel(selected_experts)
+    sorted_selected_experts = jnp.argsort(flatten_selected_experts)
+    repeat_hidden_states = jnp.repeat(inputs_2d, self.num_experts_per_tok, axis=0)
+    sorted_output = jnp.take(repeat_hidden_states, indices=sorted_selected_experts, axis=0).astype(self.dtype)
+    expert_order, expert_size = jnp.unique(flatten_selected_experts, return_counts=True, size=self.num_experts)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:337]: That's nice! I haven't tried this bincount yet. Let me see if it's jit-compatible.

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh (hunk 1 above)
Comment (RissyRan) [reply end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh:38]: Based on previous PR, the test was not passing near 0.1. I ran some [exp](https://docs.google.com/spreadsheets/d/1o2RhkRpPXL5WD7FYb9kgm64qjuifOt0E0nfe3dc0YZE/edit?resourcekey=0-HGJ3wSqnzfMQ52O_Pf10sQ&gid=1958907274#gid=1958907274) before, and when number of layers increases, the diverge is huge.

File: MaxText/layers/linears.py (hunk 4 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:329]: Thanks Matt! I was following the [paper](https://screenshot.googleplex.com/8dwYGgsy8WE6NDL) to name the method. Let me add some comments, and hopefully it will have less confusing?

File: MaxText/layers/linears.py [hunk 7]
```diff
@@ -315,6 +326,69 @@ def generate_kernels(self, num_experts, base_emb_dim, mlp_dim):
     wo_kernel = jnp.asarray(wo_kernel, self.dtype)
     return w0_kernel, w1_kernel, wo_kernel
 
+  def permute(self, inputs, gate_logits, emb_dim):
+    inputs_2d = jnp.reshape(inputs, (-1, emb_dim))
+    weights, selected_experts = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
+    weights = jax.nn.softmax(weights.astype(self.weight_dtype), axis=-1).astype(self.dtype)
+    flatten_selected_experts = jnp.ravel(selected_experts)
+    sorted_selected_experts = jnp.argsort(flatten_selected_experts)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:334]: We need to sort the experts to fit gmm API. So inputs will be sorted from exp 0 to exp 7.

File: MaxText/layers/linears.py (hunk 3 above)
Comment (gobbleturk) [reply MaxText/layers/linears.py:357]: Even for single host we have generally have multiple devices (e.g. v4-8 is single host but has 4 devices that by default use FSDP in maxtext)

File: MaxText/layers/linears.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py]: Have a quick exp with jnp.bincount(), and did not notice issues yet. Updated to it. Thanks Matt!

File: MaxText/layers/linears.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers/linears.py:337]: Oh I misunderstood based on the name of this variable. Could we name it indeces_to_sort_by_expert?

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh (hunk 1 above)
Comment (gobbleturk) [reply end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh:38]: Hmm that is quite concerning. Have you tried to run this on a CPU? I think @khatwanimohit could help you try this to see if it is a numeric issue. I think these large atol and rtol allows many wrong implementations to pass this test, which means this isn't a very useful test

**Review by gobbleturk — CHANGES_REQUESTED** [review]

File: MaxText/layers/linears.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:337]: Sure thing. Will update.

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh (hunk 1 above)
Comment (RissyRan) [reply end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh:38]: Yes, I tried on CPU in that sheet. Discussing this offline.

**Review by gobbleturk — APPROVED** [review]
Chatted with Ran offline, although the logits differ for increasing layer count, inference still produces reasonable results.
